geth/peers/peerpool.go

```diff
@@ -0,0 +1,148 @@
+	ErrDiscv5NotRunning = errors.New("Discovery v5 is not running")
+)
+
+const (
+	// expirationPeriod is a amount of time while peer is considered as a connectable
+	expirationPeriod = 60 * time.Minute
+	// DefaultFastSync is a recommended value for aggressive peers search.
+	DefaultFastSync = 3 * time.Second
+	// DefaultSlowSync is a recommended value for slow (background) peers search.
+	DefaultSlowSync = 30 * time.Minute
+)
+
+// NewPeerPool creates instance of PeerPool
+func NewPeerPool(config map[discv5.Topic]params.Limits, fastSync, slowSync time.Duration, cache *Cache, stopOnMax bool) *PeerPool {
+	return &PeerPool{
+		config:    config,
+		fastSync:  fastSync,
+		slowSync:  slowSync,
+		cache:     cache,
+		stopOnMax: stopOnMax,
+	}
+}
+
+type peerInfo struct {
+	// discoveredTime last time when node was found by v5
+	discoveredTime mclock.AbsTime
+	// connected is true if node is added as a static peer
+	connected bool
+
+	node *discv5.Node
+}
+
+// PeerPool manages discovered peers and connects them to p2p server
+type PeerPool struct {
+	// config can be set only once per pool life cycle
+	config    map[discv5.Topic]params.Limits
+	fastSync  time.Duration
+	slowSync  time.Duration
+	cache     *Cache
+	stopOnMax bool
+
+	mu                 sync.RWMutex
+	topics             []*TopicPool
+	serverSubscription event.Subscription
+	quit               chan struct{}
+
+	wg sync.WaitGroup
+}
+
+// Start creates discovery search query for each topic and consumes peers found in that topic
+// in separate loop.
+func (p *PeerPool) Start(server *p2p.Server) error {
+	if server.DiscV5 == nil {
+		return ErrDiscv5NotRunning
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.quit = make(chan struct{})
+	p.topics = make([]*TopicPool, 0, len(p.config))
+	for topic, limits := range p.config {
+		topicPool := NewTopicPool(topic, limits, p.slowSync, p.fastSync)
+		if err := topicPool.StartSearch(server); err != nil {
+			return err
+		}
+		p.topics = append(p.topics, topicPool)
+	}
+
+	events := make(chan *p2p.PeerEvent, 20)
+	p.serverSubscription = server.SubscribeEvents(events)
+	p.wg.Add(1)
+	go func() {
+		p.handleServerPeers(server, events)
+		p.wg.Done()
+	}()
+	return nil
+}
+
+func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.PeerEvent) {
+	for {
+		select {
+		case <-p.quit:
+			return
+		case event := <-events:
+			if event.Type == p2p.PeerEventTypeDrop {
+				p.mu.Lock()
+				for _, t := range p.topics {
+					t.ConfirmDropped(server, event.Peer, event.Error)
+					// restart discv5 if peers number dropped too low
+				}
+				p.mu.Unlock()
+			} else if event.Type == p2p.PeerEventTypeAdd {
+				p.mu.Lock()
+				total := 0
+				for _, t := range p.topics {
+					t.ConfirmAdded(server, event.Peer)
+					if p.stopOnMax && t.MaxReached() {
+						total++
+						t.StopSearch()
+					}
+				}
+				if p.stopOnMax && total == len(p.config) {
+					log.Debug("closing discv5 connection")
+					server.DiscV5.Close()
```

Comment:
probably it should, i decided to add that code in a separate PR https://github.com/status-im/status-go/pull/801

i am not sure if we want to have it. as i understand mobile node doesn't live too long, so it may be actually simpler just to require full node restart

Reply:
We can leave it for now. I assume that #801 contains also commits from this PR so let's focus on merging this one first.